book/internals/setters.nim

```diff
@@ -0,0 +1,76 @@
+import nimib, nimibook
+
+nbInit(theme = useNimibook)
+
+nbText: """
+## **Setters**
+Setters are a convenience mechanism for widgets.
+Through them you can hide fields on your `WidgetState`.
+Parent-widgets will be able to assign to "pseudo-fields", whose assigned values get converted and then forwarded to the actual field on `WidgetState`.
+
+allow parent-widgets to assign to fields on `WidgetState` with different types of values that get converted to the field on `WidgetState`!
+
+Every setter consists of:
+- a `setter <PseudoFieldName>` declaration with the type that the setter receives, e.g. `setter margins: int` defines a setter that will receive an int from a parent-widget
+- a `has<PseudoFieldName>=` proc, forwarding booleans that would enable/disable the pseudo-field to the actual field on `WidgetState`
+- a `val<PseudoFieldName>=` proc, converting the value received by the pseudo-field as needed and assigning it to the actual field on `WidgetState`
+
+Note: You only need 1 `has<PseudoFieldName>` proc, even if you define multiple setters.
+
+Lets take a look at a code example:
+"""
+
+nbCode:
+  ## The custom widget
+  import owlkettle 
+  
+  viewable MyViewable:
+    internalMargins: array[3, int]
+
+    setter margins: int
+    setter margins: array[3, int] 
+
+  proc `hasMargins=`*(widget: MyViewable, has: bool) =
+    widget.hasInternalMargins = has
+
+  proc `valMargins=`*(widget: MyViewable, margin: int) =
+    widget.valInternalMargins = [margin, margin, margin]
+
+  proc `valMargins=`*(widget: MyViewable, margins: array[3, int]) =
+    widget.valInternalMargins = margins
+
+  method view(state: MyViewableState): Widget =
+    gui:
+      Box():
+        Box(orient = OrientY) {.vAlign: AlignStart.}:
+          Label(text = "Text1", margin = Margin(bottom: state.internalMargins[0]))
+          Label(text = "Text2", margin = Margin(bottom: state.internalMargins[0]))
+          Label(text = "Text3", margin = Margin(bottom: state.internalMargins[0]))
+
+  ## The App
+  viewable App:
+    discard
+
+  method view(app: AppState): Widget =
+    result = gui:
+      Window:
+        Box(orient = OrientX):
+          MyViewable(margins = [0, 60, 60])
+          MyViewable(margins = 40)
+
+  when not defined(owlkettleNimiDocs):
+    brew(gui(App()))
+
+nbText: """
+This defines a `MyViewable` that has a pseudo-field called `margins`, which hides the "true" field `internalMargins`.
```

Comment:
```suggestion
This defines a `MyViewable` that has a setter called `margins`, which hides the internal field `internalMargins`.
```

Reply:
I agree with your statement that "hiding" is misleading in general. Setters in that sense don't forbid direct assignment to the field, though imo it is strongly implied that using a setter means you should use the setter instead of direct assignment to the field.

Anyway, I wouldn't start talking about "hiding" in general there, so I just rephrased the entire section